File: Java/src/hashing/problems/SubarrayWithGivenSum.java
package hashing.problems;

import java.util.HashMap;

// https://practice.geeksforgeeks.org/batch/cp-47/track/DSASP-Hashing/problem/subarray-range-with-given-sum0128
public class SubarrayWithGivenSum {

	public static int subArraySum(int[] arr, int n, int sum) {
		HashMap<Integer, Integer> hashMap = new HashMap<> ( );
		int count = 0;
		int prefixSum = 0;
		for (int integer : arr) {
			prefixSum += integer;
			if (prefixSum == sum) count++;
			if (hashMap.containsKey ( prefixSum - sum ))
				count += hashMap.get ( prefixSum - sum );
			hashMap.put ( prefixSum, hashMap.getOrDefault ( prefixSum, 0 ) + 1 );
		}
		return count;
	}

	public static void main(String[] args) {
		System.out.println ( subArraySum ( new int[]{10, 2, -2, -20, 10}, 5, -10 ) );   // 3
		System.out.println ( subArraySum ( new int[]{1, 4, 20, 3, 10, 5}, 6, 33 ) );    // 1
		System.out.println ( subArraySum ( new int[]{24, 16, 5, 26, 27, 2}, 6, 2 ) );    // 1
	}
}

File: Python/SubarrayWithGivenSum.py
# https://practice.geeksforgeeks.org/problems/subarray-with-given-sum-1587115621/1

#Function to find a continuous sub-array which adds up to a given number.
class Solution:
    def subArraySum(self,arr, n, Sum):
        Map = {}
        curr_sum = 0
        for i in range(0, n):
            curr_sum = curr_sum + arr[i]
            if curr_sum == Sum:
                return [1,i+1]
            if (curr_sum - Sum) in Map:
                return [Map[curr_sum - Sum]+2,i+1]
            Map[curr_sum] = i
        return [-1]